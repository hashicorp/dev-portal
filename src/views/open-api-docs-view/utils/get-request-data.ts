/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

// Utils
import {
	getPropertyDetailPropsFromParameter,
	getPropertyDetailPropsFromSchemaObject,
} from './get-property-detail-props'
// Types
import type { PropertyDetailsGroup } from '../components/operation-details'
import type { PropertyDetailsProps } from '../components/property-details'
import type { OpenAPIV3 } from 'openapi-types'

/**
 * Given OpenAPI parameter objects,
 * Return request data formatted for display.
 */
export async function getRequestData(
	parameters: (OpenAPIV3.ParameterObject | OpenAPIV3.ReferenceObject)[],
	requestBody?: OpenAPIV3.RequestBodyObject | OpenAPIV3.ReferenceObject
): Promise<PropertyDetailsGroup[]> {
	// Build arrays of path, query, and body parameters
	const pathParameters: PropertyDetailsProps[] = []
	const queryParameters: PropertyDetailsProps[] = []
	if (Array.isArray(parameters)) {
		for (const parameter of parameters) {
			// Skip references
			if ('$ref' in parameter) {
				continue
			}
			// Skip parameters that are not "in" the request
			if (!('in' in parameter)) {
				continue
			}
			// Parse parameters by type
			if (parameter.in === 'path') {
				pathParameters.push(
					await getPropertyDetailPropsFromParameter(parameter)
				)
			} else if (parameter.in === 'query') {
				queryParameters.push(
					await getPropertyDetailPropsFromParameter(parameter)
				)
			}
		}
	}
	// Build body parameters from requestBody data, if present
	const bodyParameters =
		requestBody && !('$ref' in requestBody)
			? await getBodyParameterProps(requestBody)
			: []
	// Build an array of request data, using any parameters present
	const requestData: PropertyDetailsGroup[] = []
	if (pathParameters.length > 0) {
		requestData.push({
			heading: 'Path Parameters',
			propertyDetails: pathParameters,
		})
	}
	if (queryParameters.length > 0) {
		requestData.push({
			heading: 'Query Parameters',
			propertyDetails: queryParameters,
		})
	}
	if (bodyParameters.length > 0) {
		requestData.push({
			heading: 'Body Parameters',
			propertyDetails: bodyParameters,
		})
	}
	// Return everything
	return requestData
}

/**
 * Given the parameter object from a body parameter,
 * which we expect to be an object with a schema if it exists,
 *
 * Return property detail data.
 */
export async function getBodyParameterProps(
	requestBody: OpenAPIV3.RequestBodyObject
): Promise<PropertyDetailsProps[]> {
	const schema = requestBody.content['application/json'].schema
	// If we don't find the expected schema, return an empty array
	if (!schema) {
		return []
	}
	// We don't expect references, but for typing purposes we handle them.
	if ('$ref' in schema) {
		return []
	}
	/**
	 * Note: we expect body schemas to always be objects at their root.
	 * We flatten the body properties to avoid showing a redundant object.
	 */
	const bodyProps = []
	const requiredProperties = schema.required || []
	for (const [key, value] of Object.entries(schema.properties)) {
		// Skip reference objects, we expect these to be de-referenced
		// before this function is called
		if ('$ref' in value) {
			continue
		}
		// Skip read-only properties
		if (value.readOnly) {
			continue
		}
		const isRequired = requiredProperties.includes(key)
		// Push props
		bodyProps.push(
			await getPropertyDetailPropsFromSchemaObject(key, value, isRequired)
		)
	}
	return bodyProps
}
