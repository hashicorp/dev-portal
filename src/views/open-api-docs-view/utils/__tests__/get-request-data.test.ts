import { getRequestData } from '../get-request-data'
import petstore3 from '../__fixtures__/petstore3.openapi'

describe('getRequestData', () => {
	it('parses non-json request bodies', async () => {
		const request = petstore3.paths['/pet/{petId}/uploadImage'].post
		const groups = await getRequestData(
			request.parameters,
			request.requestBody,
			'test-prefix'
		)
		expect(groups).toMatchSnapshot()
	})
})
