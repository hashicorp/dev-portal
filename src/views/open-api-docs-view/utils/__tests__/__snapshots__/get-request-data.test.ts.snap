// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getRequestData > parses non-json request bodies 1`] = `
[
  {
    "heading": {
      "slug": "test-prefix_path-parameters",
      "text": "Path Parameters",
    },
    "propertyDetails": [
      {
        "description": "<p>ID of pet to update</p>",
        "isRequired": true,
        "name": "petId",
        "nestedProperties": [],
        "slug": "test-prefix_path-parameters_petId",
        "type": "integer",
      },
    ],
  },
  {
    "heading": {
      "slug": "test-prefix_query-parameters",
      "text": "Query Parameters",
    },
    "propertyDetails": [
      {
        "description": "<p>Additional Metadata</p>",
        "isRequired": false,
        "name": "additionalMetadata",
        "nestedProperties": [],
        "slug": "test-prefix_query-parameters_additionalMetadata",
        "type": "string",
      },
    ],
  },
  {
    "heading": {
      "slug": "test-prefix_body-parameters",
      "text": "Body Parameters",
    },
    "propertyDetails": [
      {
        "description": "binary",
        "isRequired": "false",
        "name": "application/octet-stream",
        "nestedProperties": [],
        "slug": "test-prefix_body-parameters",
        "type": "string",
      },
    ],
  },
]
`;
